test/unit/mocks.py

```diff
@@ -0,0 +1,66 @@
+import subprocess
+
+
+class Run:
+    """Mock for subprocess.run.
+
+    When subprocess.run is called, the mock returns a subprocess.CompletedProcess
+    instance with data passed to :meth:`handle` for those command-line arguments.
+    Or, if returncode is nonzero, it raises a subprocess.CalledProcessError.
+
+    This also asserts that the correct keyword args are passed to subprocess.run,
+    for example check=True.
+    """
+
+    def __init__(self):
+        self._commands = {}
+        self.call_count = 0
+
+    def handle(self, args: list[str], returncode: int = 0, stdout: str = '', stderr: str = ''):
+        """Handle specified command-line args with the given return code, stdout, and stderr."""
+        self._commands[tuple(args)] = (returncode, stdout, stderr)
```

Comment:
I like how explicit this is. I wonder if it would be nice as a context manager so that it's linked to the specific calls that use `subprocess.run` rather than just the test function. Probably fine as is though, I don't think any of the tests ever even call `handle` more than once. But just for fun ...
```suggestion
    @contextmanager
    def handle(self, args: list[str], returncode: int = 0, stdout: str = '', stderr: str = '') -> None:
        """Handle specified command-line args with the given return code, stdout, and stderr."""
        key = tuple(args)
        assert key not in self._commands
        self._commands[key] = (returncode, stdout, stderr)
        yield None
        del self._commands[key]
```
Used like:
```py
with run.handle(['foo', 'bar']):
    thing.foo('bar')
with (
    run.handle(['foo', 'bar']),
    run.handle(['foo' '--cleanup'], returncode=9000),
    pytest.raises(CalledProcessError) as excinfo,
):
    thing.foo('bar', cleanup=True)
assert run.call_count == 3
```

Reply:
Yeah, I don't think this is necessary here (if we were exporting it as a more general mock, maybe), and it's better as a test-level pytest fixture, as it fits with pytest and avoids the indentation. You can actually call `handle` multiple times in the test if you want, to update it (you could even call it in wait's `ready` func if you wanted to get clever, but I didn't need to).

Reply:
I agree that it's not necessary here. I just want to clarify that I think you'd still use the version I suggested as a pytest fixture rather than (e.g.) doing the monkeypatching in the context manager, as that forces you to handle *all* calls to `run` (like you do currently). I just like the idea of specific `handle` setups only existing as long as needed, and automatically un-handling themselves so that if you do need to handle multiple calls in one test, it's clear which call you're expecting where, if that makes sense. But yeah, agreed that it's not necessary here.